### PR TITLE
Add support for Rust edition

### PR DESCRIPTION
**Pull request by kpp**
Books will have to replace ` ```rust` with ` ```edition2018` everywhere to make it work, but it would be useful to change it in config only.

According to https://github.com/rust-lang/mdBook/issues/812#issuecomment-554758875

### CODE REVIEW

**Comment by ehuss** [comment]
Oh, thanks for pushing this forward!

My preference would be to have a new top-level `[rust]` table where rust-specific configuration can go.  We can add more things to it in the future (maybe even things like dependencies).  

As for forwarding the configuration into the playground, it shouldn't be too hard.  I think it would involve adding another field to `RenderItemContext` in hbs_renderer so the relevant information can make it into `add_playpen_pre`.

**Comment by kpp** [comment]
I feel bugs in the future. ` ```rust` blocks are replaced with `class="language-rust"`. How about pure ` ```edition2018` code blocks?

File: src/renderer/html_handlebars/hbs_renderer.rs [hunk 1]
```diff
@@ -612,10 +619,24 @@ fn add_playpen_pre(html: &str, playpen_config: &Playpen) -> String {
                 if (!classes.contains("ignore") && !classes.contains("noplaypen"))
                     || classes.contains("mdbook-runnable")
                 {
+                    let contains_e2015 = classes.contains("edition2015");
+                    let contains_e2018 = classes.contains("edition2018");
+                    let edition_class = if contains_e2015 || contains_e2018 {
+                        // the user forced edition, we should not overwrite it
```
Comment (kpp) [inline src/renderer/html_handlebars/hbs_renderer.rs]: This allow users to write ` ```rust,edition2015` even with book.edition=2018 and their code will be sent to playpen with edition2015. I don't know how to implement the same behavior in `mdbook test`.

Comment (ehuss) [reply src/renderer/html_handlebars/hbs_renderer.rs]: `mdbook test` with a nightly toolchain honors the `edition2015` annotation.  1.41 will honor it as of https://github.com/rust-lang/rust/pull/66238.

**Comment by ehuss** [comment]
> I feel bugs in the future. `` ```rust `` blocks are replaced with `class="language-rust"`. How about pure `` ```edition2018 `` code blocks?

Hm, if I understand the question correctly:  rustdoc treats code blocks with annotations like `edition2018` or `no_run` as rust blocks, but mdbook does not render them as rust blocks. Is that what you are referring to?

That sounds like a separate issue, and something we can maybe deal with separately.

File: src/renderer/html_handlebars/hbs_renderer.rs (hunk 1 above)
Comment (kpp) [reply src/renderer/html_handlebars/hbs_renderer.rs]: Will it resolve in an error if there will be `edition=2015` in bool.toml and ` ```edition2018` in the block of code?

Comment (ehuss) [reply src/renderer/html_handlebars/hbs_renderer.rs]: No, the inline annotation overrides the global default.

**Comment by kpp** [comment]
> That sounds like a separate issue

I agree.

**Comment by ehuss** [comment]
Can you also move it under a separate `[rust]` table?

**Comment by kpp** [comment]
It will take some time. I am not sure I will do that in a few days.

**Comment by ehuss** [comment]
Completed via #1163.

**Pull request closed without merging**